Fix #7016: Separate review test to pass/fail mode and add score bar for each skill

sophiewu6 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
Fix #7016: This PR aims to pass or fail users depending on their review test scores and added a new card containing score bar for each skill.
![Screen Shot 2019-06-29 at 5 03 30 PM](https://user-images.githubusercontent.com/32754302/60390760-8ba10880-9a92-11e9-98f3-81ecc7b30a59.png)
pass mode 
<img width="842" alt="Screen Shot 2019-07-01 at 10 40 56 AM" src="https://user-images.githubusercontent.com/32754302/60456399-07639800-9bee-11e9-943f-224f6c01e8d2.png">
fail mode


## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

vinitamurthi committed:
Draft PR that adds a question player directive

vinitamurthi committed:
Use the Question Player Backend API to fetch questions when starting the question player directive

vinitamurthi committed:
Fix linter error in StateTranslationEditorDirective.js

vinitamurthi committed:
Merge remote-tracking branch 'upstream/develop' into question_ui

vinitamurthi committed:
Merge remote-tracking branch 'upstream/develop' into question_ui

vinitamurthi committed:
Merge remote-tracking branch 'upstream/develop' into question_body

vinitamurthi committed:
Initial check in for question body

vinitamurthi committed:
Update question progression

vinitamurthi committed:
Stop getExplorationID from failing when ins question player mode

vinitamurthi committed:
Disable new structures players

vinitamurthi committed:
Fixed linter errors

vinitamurthi committed:
More linter fixes

vinitamurthi committed:
Merge develop and fix conflict

vinitamurthi committed:
Merge remote-tracking branch 'upstream/develop' into question_body

vinitamurthi committed:
Fix merge conflicts

vinitamurthi committed:
Fixed linter error

vinitamurthi committed:
Fix merge conflicts

vinitamurthi committed:
Removed changes of empty lines

vinitamurthi committed:
Added exploration footer to the question player directive

vinitamurthi committed:
Fix linter errors

vinitamurthi committed:
More linter error fixes

vinitamurthi committed:
even more linter fixes

Commit:
Merge remote-tracking branch 'upstream/develop' into question_body

vinitamurthi committed:
Add question player backend dependency in exploration editor and player pages, this is to avoid an injector issue

vinitamurthi committed:
Initial check in for score calculation

vinitamurthi committed:
Fixed merge conflicts

vinitamurthi committed:
Fix linter errors

vinitamurthi committed:
More linter fixes

vinitamurthi committed:
Linter fixes

vinitamurthi committed:
Linter fix

vinitamurthi committed:
LInter fix

vinitamurthi committed:
Linter fixes

vinitamurthi committed:
More linter fixes

vinitamurthi committed:
LInter fix

vinitamurthi committed:
linter fix

vinitamurthi committed:
Linter

vinitamurthi committed:
Saving score calc state

vinitamurthi committed:
Linter fixes

vinitamurthi committed:
Calculate total score

vinitamurthi committed:
Merged with upstream

vinitamurthi committed:
Result player action buttons, also fixed correctness footer alignment

vinitamurthi committed:
CSS Fixes

vinitamurthi committed:
Fixed merge conflicts

vinitamurthi committed:
Disabled new structure flags

vinitamurthi committed:
Linter fixes

vinitamurthi committed:
Fix linter issue

vinitamurthi committed:
Add package-lock

vinitamurthi committed:
Fixed merge conflicts

vinitamurthi committed:
fetch skill summaries

vinitamurthi committed:
Merge upstream/develop to branch

vinitamurthi committed:
LInt fix

vinitamurthi committed:
Remove duplicate mention of service

vinitamurthi committed:
Lint fix

vinitamurthi committed:
Merge upstream/develop

vinitamurthi committed:
Remove merge conflict indication in practice_session.py

vinitamurthi committed:
Added file overview

vinitamurthi committed:
Remove constants

vinitamurthi committed:
Remove unnecessary changes

vinitamurthi committed:
Lint fix

vinitamurthi committed:
Remove changes in practice_sessions.py

vinitamurthi committed:
Remove extra line in question model

vinitamurthi committed:
Merge remote-tracking branch 'upstream/develop' into score_calc

vinitamurthi committed:
Pass skill list to question player config

vinitamurthi committed:
Calculate scores per skill

vinitamurthi committed:
Lint fix

vinitamurthi committed:
Enable review tests for the results page

vinitamurthi committed:
Test coverage and review changes

vinitamurthi committed:
Merge remote-tracking branch 'upstream/develop' into score_calc

vinitamurthi committed:
Lint fix

vinitamurthi committed:
Remove debug statements

vinitamurthi committed:
Remove trailing space

vinitamurthi committed:
Fix test coverage and review comments

sophiewu6 committed:
Merge branch 'score_calc' of https://github.com/vinitamurthi/oppia into temp

vinitamurthi committed:
Fix review comments

vinitamurthi committed:
Remove totalScore variable

vinitamurthi committed:
Fix review test UI and rename some CSS classes

vinitamurthi committed:
Rename skills_with_description to skill_descriptions

vinitamurthi committed:
Merge remote-tracking branch 'upstream/develop' into score_calc

vinitamurthi committed:
Fixed merge conflicts

sophiewu6 committed:
pass or fail tests

sophiewu6 committed:
merge conflicts

sophiewu6 committed:
lint

vinitamurthi commented:
Can you fix the merge conflicts and push again? I'll take a first pass until then.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -1,56 +1,224 @@
+          </th>
+        </tr>
+        <tr ng-repeat="scorePerSkill in $ctrl.scorePerSkillMapping">
+          <td> {{scorePerSkill.description}} </td>
```
Replace {{scorePerSkill.description}} with <[scorePerSkill.description]>

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -1,56 +1,224 @@
+        <tr ng-repeat="scorePerSkill in $ctrl.scorePerSkillMapping">
+          <td> {{scorePerSkill.description}} </td>
+          <td>
+            <div class="oppia-score-per-skill">
```
Change class name to skill-score-container or something on those lines?

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -1,56 +1,224 @@
+          </td>
+        </tr>
+      </table>
+      <button class="btn btn-default" ng-if="!$ctrl.passTest"> Review and Retry </button>
```
change 'passTest' to 'hasUserPassedTest' or 'isTestPassed' or something like that

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -1,56 +1,224 @@
+    margin-bottom: 16px;
+  }
+
+  .skill-score-table .skill-score-table-headings {
```
Add the parent class as well (i.e.: oppia-question-player-container) before all the CSS classes to ensure we only affect the question player classes.

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -118,14 +122,95 @@ oppia.directive('questionPlayer', [
           ctrl.currentProgress = 0;
+          ctrl.totalScore = 0.0;
+          ctrl.scorePerSkillMapping = {};
+          ctrl.passTest = true;
```
Change to isTestPassed/hasUserPassedTest or something

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
           };
 
+          var isInPassOrFailMode = function() {
+            return UrlService.getPathname().includes('review_test');
```
Let's not do it this way. I wan't to avoid using URLs to understand what to do, instead we should create a configuration value for the question player.
Please pass a new parameter 'questionPlayerMode' in the questionPlayerConfig and set the following values:
`questionPlayerMode: {
      modeType: 'PASS_FAIL', 
      passCutoff: 0.75
}`
Then instead of this line you can say something like 
`return (questionPlayerConfig.questionPlayerMode && questionPlayerConfig.questionPlayerMode.modeType ==='PASS_FAIL'`

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
+            $scope.resultsLoaded = true;
+          };
+
+          var passOrFailTests = function() {
```
Change function name to hasUserPassedTest, and return the boolean value instead of setting it here. Set it below when you call the function.

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
+              });
+            }
+
+            if (!ctrl.passTest) {
```
Why are you doing this?

sophiewu6 replied:
Sorry I meant to clear action buttons in failing mode. It was a typo. I have fixed it.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
+            if (resultHashString) {
+              var questionStateData = JSON.parse(resultHashString);
+              calculateScores(questionStateData);
+              passOrFailTests();
```
Change to `$ctrl.isTestPassed = hasUserPassedTest();`

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
+
+          var passOrFailTests = function() {
+            ctrl.passTest = true;
+            if (isInPassOrFailMode()) {
```
Change the following statement to:

```
if (!isInPassOrFailMode()) { 
   return true;
}
```

sophiewu6 replied:
I chose to return at the end. Is that okay?

vinitamurthi replied:
Yep that's fine!

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
+                }
+              });
+            }
+
```
Return the passTest value instead of setting it here.

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
+
+          ctrl.getColorForScore = function(scorePerSkill) {
+            var correctionRate = scorePerSkill.score / scorePerSkill.total;
+            if (correctionRate >= CORRECTION_RATE_TO_PASS) {
```
Don't use this constant, have it passed in questionPlayerConfig (See my above comment about the questionPlayerMode)

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -154,12 +239,130 @@ oppia.directive('questionPlayer', [
+          ctrl.getColorForScore = function(scorePerSkill) {
+            var correctionRate = scorePerSkill.score / scorePerSkill.total;
+            if (correctionRate >= CORRECTION_RATE_TO_PASS) {
+              return 'rgb(0, 150, 136)';
```
Can you mention the color name (i.e. red/orange/yellow) in comments here? Just so that it's easier to recognize

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/pages/review-test-page/review-test-page.directive.ts:
```diff
@@ -56,11 +60,43 @@ oppia.directive('reviewTestPage', ['UrlInterpolationService', function(
+                  url: storyViewerUrl
+                }
+              ],
+              skillList: skillIdList,
```
Add questionPlayerMode to this config. (See above comment of the values for questionPlayerMode)

sophiewu6 replied:
Done.

vinitamurthi commented:
@sophiewu6  One more thing - I am also not sure about keeping the passed skill color as green since the score ring has a red color (which is what you are using for failed skills).
Can you do one thing? Make the score ring also green in color if the user passes, and make score ring and failed skills as the red color if the user fails the test

sophiewu6 committed:
merge conflicts

sophiewu6 committed:
review changes

sophiewu6 committed:
lint

sophiewu6 committed:
lint

sophiewu6 commented:
Hi @vinitamurthi I have fixed the merge conflicts and resolved all review comments. I also updated the screenshots so that you could check the new score ring color. PTAL!

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.constants.ts:
```diff
@@ -15,5 +15,6 @@
 /**
  * @fileoverview Constants for the question player directive.
  */
+
```
Remove this line

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -17,13 +17,43 @@ <h3 class="oppia-question-player-header">Question <[$ctrl.currentQuestion]>/<[$c
-      <score-ring score="$ctrl.totalScore"></score-ring>
-      <h4 class="learn-more"> Learn more about your score </h4>
+      <h3 class="oppia-question-player-header" ng-if="$ctrl.isTestPassed"> Test Complete. Well Done! </h3>
+      <h3 class="oppia-question-player-header" ng-if="!$ctrl.isTestPassed"> Test Failed. </h3>
```
I am wondering if there's a better way to indicate that a test has failed. What do you think of something like 'Test Failed. Let's Review'
Do you have any ideas?

sophiewu6 replied:
Changed it to 'Test failed. Please Review And Retry.' Does this sound good?

vinitamurthi replied:
Okay that's fine !

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -17,13 +17,43 @@ <h3 class="oppia-question-player-header">Question <[$ctrl.currentQuestion]>/<[$c
+        </colgroup>
+        <tr>
+          <th class="skill-score-table-headings">
+            <p> Skill Descriptions </p>
```
I don't think you need the` <p>` tag, you can leave it as is or use <span> instead. Then you won't need the `display: inline` CSS style for this below.

sophiewu6 replied:
Removed.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -153,6 +183,39 @@ <h4 class="learn-more"> Learn more about your score </h4>
+    text-align: center;
+  }
+
+  .oppia-question-player-container .skill-score-table .skill-score-table-headings p {
```
See above comment about removing the paragraph and either keeping it empty or using a `<span>` tag. You can remove this style

sophiewu6 replied:
Removed.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -105,6 +105,7 @@ require('domain/question/QuestionBackendApiService.ts');
 require('domain/utilities/UrlInterpolationService.ts');
 
 require('pages/interaction-specs.constants.ts');
+require('services/contextual/UrlService.ts');
```
Remove UrlService now that you aren't using it anymore.

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -124,19 +125,21 @@ oppia.directive('questionPlayer', [
         '$scope', '$sce', '$rootScope', '$location',
         '$sanitize', '$window', 'HtmlEscaperService',
-        'QuestionPlayerBackendApiService',
+        'QuestionBackendApiService', 'UrlService',
```
Remove UrlService

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -124,19 +125,21 @@ oppia.directive('questionPlayer', [
             $scope, $sce, $rootScope, $location,
             $sanitize, $window, HtmlEscaperService,
-            QuestionPlayerBackendApiService) {
+            QuestionBackendApiService, UrlService) {
```
Remove UrlService

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -282,17 +294,53 @@ oppia.directive('questionPlayer', [
+
+          ctrl.getColorForScore = function(scorePerSkill) {
+            var correctionRate = scorePerSkill.score / scorePerSkill.total;
+            if (correctionRate >=
```
So suppose questionPlayerMode doesn't exist, this function will fail.
To account for that, let's always return the green color if questionPlayerMode doesn't exist. Then do all this logic in case questionPlayerMode does exist.

sophiewu6 replied:
Done.

vinitamurthi commented on core/templates/dev/head/pages/review-test-page/review-test-page.directive.ts:
```diff
@@ -24,6 +24,7 @@ require(
   'question-player.directive.ts');
 require('interactions/interactionsQuestionsRequires.ts');
 require('objects/objectComponentsRequiresForPlayers.ts');
+
```
Remove extra line

sophiewu6 replied:
Done.

vinitamurthi reviewed (changes requested):
Looks good! Just a few things

sophiewu6 committed:
review changes

vinitamurthi reviewed (approved):
Looks good! Only one last thing - the 'Review and Retry' button isn't very prominent. Can you make it a blue background (Use color: #0D48A1) with white text ? Also tests seem to be failing but otherwise LGTM!

sophiewu6 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into pass-or-fail-tests

sophiewu6 committed:
color change

sophiewu6 commented:
@vinitamurthi I changed the color to green instead to match the background color. You could see it in the screenshot above.

vinitamurthi commented:
> @vinitamurthi I changed the color to green instead to match the background color. You could see it in the screenshot above.

All of our buttons so far have been in that blue color I mentioned above, so better to set it to blue I think!

sophiewu6 commented:
@vinitamurthi Above the result page, there are two buttons "DONATE" and "CREATE", which also have color green. I tried blue as well but it didn't look good.

vinitamurthi commented:
The integration test is failing but I don't think its related to this CL. The test failed to start itself and that's the error that came up. @oppia/dev-workflow-team any ideas? I have restarted that particular test for now
Meanwhile @nithusha21 @aks681 Can you take a look?

vinitamurthi commented:
> @vinitamurthi Above the result page, there are two buttons "DONATE" and "CREATE", which also have color green. I tried blue as well but it didn't look good.

Alright let's keep it green for now.

nithusha21 commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.html:
```diff
@@ -17,13 +17,47 @@ <h3 class="oppia-question-player-header">Question <[$ctrl.currentQuestion]>/<[$c
-      <h3 class="oppia-question-player-header"> Test Complete. Well Done! </h3>
-      <score-ring score="$ctrl.totalScore"></score-ring>
-      <h4 class="learn-more"> Learn more about your score </h4>
+      <h3 class="oppia-question-player-header" ng-if="$ctrl.isTestPassed"> Test Complete. Well Done! </h3>
```
Names that start with "is" usually refer to a function name. Please follow this for booleans.

```suggestion
      <h3 class="oppia-question-player-header" ng-if="$ctrl.testIsPassed"> Test Complete. Well Done! </h3>
```

sophiewu6 replied:
Done.

nithusha21 commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -236,23 +238,32 @@ oppia.directive('questionPlayer', [
+          var isInPassOrFailMode = function() {
+            return (ctrl.questionPlayerConfig.questionPlayerMode &&
+              ctrl.questionPlayerConfig.questionPlayerMode.modeType ===
+              'PASS_FAIL');
```
Maybe better to define this as a constant somewhere and use it throughout.

sophiewu6 replied:
Done.

nithusha21 commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -284,17 +295,57 @@ oppia.directive('questionPlayer', [
           };
 
+          var hasUserPassedTest = function() {
+            var isTestPassed = true;
```
rename to testIsPassed

sophiewu6 replied:
Done.

nithusha21 commented on core/templates/dev/head/components/question-directives/question-player/question-player.directive.ts:
```diff
@@ -284,17 +295,57 @@ oppia.directive('questionPlayer', [
+          ctrl.getColorForScore = function(scorePerSkill) {
+            if (!isInPassOrFailMode()) {
+              // return color green when not in pass or fail mode.
+              return 'rgb(0, 150, 136)';
```
I see this used in multiple places. Can we define some standard colors and use the constants here?

sophiewu6 replied:
Done.

nithusha21 commented on core/templates/dev/head/components/score-ring/score-ring.directive.ts:
```diff
@@ -49,6 +50,26 @@ oppia.directive('scoreRing', [
+          ctrl.getScoreRingColor = function() {
+            if (ctrl.isTestPassed()) {
+              // return color green when passed.
+              return 'rgb(0, 150, 136)';
```
Don't repeat the same rgb strings. Use constants.

sophiewu6 replied:
Done.

nithusha21 commented on core/templates/dev/head/components/score-ring/score-ring.directive.ts:
```diff
@@ -49,6 +50,26 @@ oppia.directive('scoreRing', [
+          ctrl.getScoreOuterRingColor = function() {
+            if (ctrl.isTestPassed()) {
+              // return color green when passed.
+              return 'rgb(143, 217, 209)';
```
Is this green too? I see 2 different RGB values commented as green.

sophiewu6 replied:
It is shallow green for the outer ring. I also created a constant for it.

nithusha21 commented on core/templates/dev/head/components/score-ring/score-ring.directive.ts:
```diff
@@ -49,6 +50,26 @@ oppia.directive('scoreRing', [
+              return 'rgb(143, 217, 209)';
+            } else {
+              // return color orange when failed.
+              return 'rgb(244, 206, 186)';
```
Ditto, is this orange too?

sophiewu6 replied:
Same as above.

nithusha21 commented on core/templates/dev/head/pages/review-test-page/review-test-page.directive.ts:
```diff
@@ -68,10 +78,30 @@ oppia.directive('reviewTestPage', ['UrlInterpolationService', function(
+                skillIdList.length),
+              questionPlayerMode: {
+                modeType: 'PASS_FAIL',
+                passCutoff: 0.75
```
This should be stored in a constant I think (if you plan to use this elsewhere too).

sophiewu6 replied:
It will be referred to as quesionPlayerMode.passCutoff elsewhere, so I don't think we need a constant for this one.

nithusha21 commented on core/templates/dev/head/pages/review-test-page/review-test-page.directive.ts:
```diff
@@ -68,10 +78,30 @@ oppia.directive('reviewTestPage', ['UrlInterpolationService', function(
+              resultActionButtons: [
+                {
+                  type: 'BOOST_SCORE',
+                  text: 'Boost My Score'
```
Do these need to be translated? As these are learner facing. If so, you should add these as I18N strings.

vinitamurthi replied:
@nithusha21 I had a question about this comment - This string will be rendered on the HTML page (See the above screenshot of where we show 'Boost My Score'). Considering that we have been writing regular English text on HTML as well, won't this get translated the same way as any other strings in the HTML file?

nithusha21 replied:
Actually no. So if you see, we have very specific strings that get translated. You'll need to use the translate="I18N_..." attribute in the tag to translate it. The files which store the translated versions of these I18N strings are regularly updated (you'll see PRs from translatewiki titled "routine update of translations"). This is how we manage to do translations of text. Now for views that are learner facing, this is crucial (not so much for creator views). So you would want to add I18N strings for all headings, text, etc on the players. Let me know if you need some more explanation. But you could just find an I18n string, and trace it back to figure out what to add and where. I think you'll need to edit a couple of JSON files to do this (the english version, and qqq.json I think).

vinitamurthi replied:
Oh I see - so what you are saying is that even in the HTML page we have to add the I18N attributes for every piece of text right?

nithusha21 replied:
Yeah. You could look at the library page, or the learner dashboard for reference.

sophiewu6 replied:
But do we need to add every I18N string for every text in this case? I don't think we are doing this for HTML files. If that's the case, we might need to do the cleanup for many files instead of only this one text.

seanlip replied:
It needs to happen for anything that's learner-facing. Creator-facing views aren't internationalized.

sophiewu6 replied:
Ok I see. I think it is an issue for question player in general then. I will create an issue and resolve it in a separate PR.

sophiewu6 replied:
https://github.com/oppia/oppia/issues/7084 issue created

nithusha21 reviewed (commented):
I don't see any file that is owned by me, but somehow I got assigned as a codeowner? I took a pass anyway, and left some generic comments! PTAL @sophiewu6. Thanks.

sophiewu6 committed:
review changes

sophiewu6 committed:
lint

sophiewu6 committed:
lint

sophiewu6 committed:
lint

sophiewu6 commented:
@aks681 Could you please take a look? Thanks!
